IQSS/dataverse
# 6936 Implement email domain based groups

**What this PR does / why we need it**:
Next to explicit groups, IP groups and Shibboleth groups, there is another valuable source of information to which association a user belongs to: a verified email.

Based on the domain part of the email, it's very easy to create groups based on affiliations. This might be handy for installations like the one from Leuven, where different universities share one instance (multi-tenancy). Our use case @ FZJ is to be able to open up for collaboration while automatically granting some rights to staff.

As this is independent from Shibboleth, this might be good for other instances like WZB (ping @RightInTwo), that create users via API.

**Which issue(s) this PR closes**:

Closes #6936 

**Special notes for your reviewer**:

This is not only introducing a new group provider, but also refactors some other places in the codebase. Some noteable places:

- Add a simpler verification check function for email addresses and make use of it. It's much easier to understand now and easier to extend (looking at #6694).
-  Extend the usage of `ExceptionMapper`s in REST API endpoints. There are a lot of places with lots of boilerplate and catch all like try-catch code. This makes the code more verbose and violates [DRY principle](https://dzone.com/articles/software-design-principles-dry-and-kiss). Obviously I didn't touch too much of other code to keep the scope of this PR as small as possible. (But didn't want to encourage the use of catch-alls any further...)

**Suggestions on how to test this**:

- Deploy, but don't yet publish the root dataverse.
- Create a new mail domain group
- Add the group as member or similar to the root dataverse
- Create a user with a group-matching mail address
- Verify user email
- Get access to the root dataverse
- Delete the permission
- Verify permission is gone

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

Nope.

**Is there a release notes update needed for this change?**:

It should be mentioned this is a new option to create groups of people.

**Additional documentation**:

Nada.
